torch_geometric/transforms/rooted_subgraphs.py

```diff
@@ -0,0 +1,230 @@
+from typing import Optional, Tuple
+
+import torch
+from torch_sparse import SparseTensor
+
+from torch_geometric.data import Data
+from torch_geometric.transforms import BaseTransform
+
+
+class RootedSubgraphData(Data):
+    r"""A data object describing a homogeneous graph together with each node's
+    rooted subgraph. It contains several additional properties that hold the
+    information of all nodes' rooted subgraphs. Assume the data represents a
+    graph with :math:`N` nodes and :math:`M` edges, also each node
+    :math:`i\in \[N\]` has a rooted subgraph with :math:`N_i` nodes and
+    :math:`M_i` edges.
+
+    Additional Properties:
+        subgraph_nodes_mapper (LongTensor): map each node in rooted subgraph
+        to a node in the original graph.
+            Size: :math:`\sum_{i=1}^{N}N_i x 1`
+        subgraph_edges_mapper (LongTensor): map each edge in rooted subgraph to
+         a edge in the original graph.
+            Size: :math:`\sum_{i=1}^{N}M_i x 1`
+        subgraph_batch: map each node in rooted subgraph to its corresponding
+        rooted subgraph index.
+            Size: :math:`\sum_{i=1}^{N}N_i x 1`
+        combined_rooted_subgraph: edge_index of a giant graph which represents
+        a stacking of all rooted subgraphs.
+            Size: :math:`2 x \sum_{i=1}^{N}M_i`
+
+    The class works as a wrapper for the data with these properties, and
+    automatically handles mini batching for them.
+
+    """
+    def __inc__(self, key, value, *args, **kwargs):
+        num_nodes = self.num_nodes
+        num_edges = self.edge_index.size(-1)
+        if 'combined_subgraph' in key:
+            return getattr(
+                self, key[:-len('combined_subgraph')] +
+                'subgraph_nodes_mapper').size(0)
+        elif 'subgraph_batch' in key:
+            # should use number of subgraphs or number of supernodes.
+            return 1 + getattr(self, key)[-1]
+        elif 'nodes_mapper' in key or 'selected_supernodes' in key:
+            return num_nodes
+        elif 'edges_mapper' in key:
+            return num_edges
+        else:
+            return super().__inc__(key, value, *args, **kwargs)
+
+    def __cat_dim__(self, key, value, *args, **kwargs):
+        if 'combined_subgraph' in key:
+            return -1
+        else:
+            return super().__cat_dim__(key, value, *args, **kwargs)
+
+    def rooted_subgraph(self, node_index: int) -> Data:
+        """
+        Returns a :obj:`Data` object representing the rooted subgraph at
+        :obj:`node_index`.
+        """
+        nodes = self.subgraph_nodes_mapper[torch.where(
+            self.subgraph_batch == node_index)]
+        return self.subgraph(nodes)
```

Comment:
Are you sure this works as expected? In particular because we hold both the original data and the subgraph data inside this object.

Reply:
Yeah good point - I wrote it when I had it in my head that we shouldn't store both 😓

Reply:
What about we provide a function here that automatically maps existing attributes to the subgraphs? This way, we can simply call it within each mini-batch and train upon it.

Reply:
Yeah sounds reasonable: so the interface would take a list of nodes and return a data object with those (rooted) subgraphs?

Reply:
I just saw your high-level comment now. I think this is a good approach we could follow. Alternatively, we can do the conversion as part of a mapping function inside `RootedSubgraphData`. That is, we leave the storage as it is due to memory constraints, and then do:
```python
for data in loader:
    data = data.map_features()  # or different naming
    model(data.x, data.edge_index, data.subgraph_batch, data.batch)
```

Reply:
right that makes sense - I guess we could also overwrite the `get_attr` to avoid the `map_features`, but this might be more confusing than not.

Reply:
After looking at it a bit more, I think I prefer your suggestion. If its designed this way, there is no reason to have the mini batching `__inc__` and `__catdim__` because the graph size doesn't actually change until `data.map_features()`

Reply:
On second thought this would require recomputing the subgraph each time. Wil go with your suggestion